wicker/core/datasets.py

```diff
@@ -125,79 +135,134 @@ def __len__(self) -> int:
         pass
 
     @abc.abstractmethod
-    def schema(self) -> DatasetSchema:
-        """Return the schema of the dataset."""
+    def arrow_table(self) -> pyarrow.Table:
+        """Return the pyarrow table with all the metadata fields and pointers of the dataset."""
         pass
 
     @abc.abstractmethod
-    def arrow_table(self) -> pyarrow.Table:
-        """Return the pyarrow table with all the metadata fields and pointers of the dataset."""
+    def schema(self) -> DatasetSchema:
+        """Return the schema of the dataset."""
         pass
 
 
```

Comment:
For the `BaseDataset` class, some of the key arguments like `pa_filesystem`, `path_factory` and `storage` should no longer be `Optional` or take an S3-specific default value. So in the `__init__` function, I just move these arguments higher in the argument list. Pardon that GitHub has a little trouble and makes the diff look more complicated than this.